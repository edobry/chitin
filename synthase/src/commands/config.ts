import { Command } from 'commander';
import { exportEnvironmentToBash } from '../shell';
import { serializeToYaml } from '../utils';
import { loadAndValidateConfig, createEnvironmentVariables } from './utils';

/**
 * Create and configure the config command
 * @returns Configured Command object
 */
export function createConfigCommand(): Command {
  const command = new Command('config')
    .description('Load and validate the Chitin configuration')
    .option('-j, --json', 'Output as JSON instead of YAML')
    .option('-e, --export-env', 'Export configuration as environment variables')
    .option('-p, --path <path>', 'Custom path to user config file')
    .action(async (options) => {
      try {
        // Load and validate configuration
        const { config, validation } = await loadAndValidateConfig({
          userConfigPath: options.path,
          exitOnError: true
        });
        
        // Output format handling
        if (options.json) {
          // Output as JSON
          console.log(JSON.stringify(config, null, 2));
        } else if (options.exportEnv) {
          // Export as environment variables
          const env = createEnvironmentVariables(config);
          
          const exportPath = await exportEnvironmentToBash(env);
          console.log(`Exported environment to ${exportPath}`);
        } else {
          // Default: Output as YAML
          console.log(serializeToYaml(config as Record<string, unknown>));
        }
      } catch (error) {
        console.error('Error loading configuration:', error instanceof Error ? error.message : String(error));
        process.exit(1);
      }
    });

  return command;
} 
