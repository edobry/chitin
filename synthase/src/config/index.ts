export * from './loader';
export * from './merger';
export * from './validator';
export * from './types'; 
