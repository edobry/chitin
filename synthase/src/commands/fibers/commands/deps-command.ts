import { Command } from 'commander';
import { serializeToYaml } from '../../../utils';
import { loadAndValidateConfig } from '../../utils';
import { discoverModulesFromConfig } from '../../../modules/discovery';
import { validateModulesAgainstConfig } from '../../../modules/validator';
import { 
  getLoadableFibers, 
  areFiberDependenciesSatisfied, 
  createChainFilter,
  getFiberIds,
  getChainIds,
  orderChainsByDependencies,
} from '../../../fiber';
import {
  orderFibers,
  getDependentFibers,
  getChainDependencies,
  ensureCoreDependencies
} from '../utils/dependency-utils';
import {
  isFiberEnabled,
  countDisplayedModules
} from '../utils/fiber-utils';
import {
  associateChainsByFiber,
  filterDisabledFibers
} from '../organization';
import { join } from 'path';
import fs from 'fs';
import yaml from 'js-yaml';
import { FIBER_NAMES } from '../../../fiber/types';
import { CONFIG_FIELDS, FILE_NAMES } from '../../../config/types';
import { EMOJI } from '../../../constants';
import { loadConfigAndModules } from '../utils/config-loader';
import { printDependencyNode } from '../utils/dependency-display';
import { DEPENDENCY_DISPLAY } from '../../../fiber/constants';

// Import generateFiberDependencyGraph from fiber/graph.ts
import { generateFiberDependencyGraph } from '../../../fiber/graph';
import { UserConfig } from '../../../config/types';
import { 
  buildFiberDependencyGraph, 
  dependencyGraphToJson,
  FiberDependencyGraph,
  FiberEnvironment
} from '../../../fiber/dependency-graph';

/**
 * Create the 'deps' subcommand for showing fiber dependencies in a diagram
 * @returns Configured Command object
 */
export function createDepsCommand(): Command {
  return new Command('deps')
    .description('Display fiber dependency diagram')
    .option('-p, --path <path>', 'Custom path to user config file')
    .option('-b, --base-dirs <dirs...>', 'Additional base directories to scan for modules')
    .option('-H, --hide-disabled', 'Hide disabled fibers')
    .option('-r, --reverse', 'Show reverse dependencies (what requires this fiber)')
    .option('-d, --detailed', 'Show detailed dependency information')
    .option('-f, --flat', 'Show flat list instead of tree view')
    .option('-g, --graphviz', 'Output dependencies in GraphViz DOT format')
    .option('-j, --json', 'Output dependency information in JSON format')
    .action(async (options) => {
      try {
        const environment = await loadConfigAndModules(options);
        const { config, moduleResult, displayFiberIds, orderedFibers } = environment;

        // Build the fiber dependency graph using our extracted utility
        const graph = buildFiberDependencyGraph(
          environment as FiberEnvironment,
          { 
            hideDisabled: options.hideDisabled, 
            reverse: options.reverse 
          }
        );
        
        // Output in JSON format if requested
        if (options.json) {
          console.log(JSON.stringify(dependencyGraphToJson(graph), null, 2));
          return;
        }
        
        // Use the appropriate map based on whether we want to show dependencies or reverse dependencies
        const { dependencyMap, reverseDependencyMap, detectionInfo, fibersToShow, rootFibers } = graph;
        const activeMap = options.reverse ? reverseDependencyMap : dependencyMap;
        
        // Output GraphViz DOT format
        if (options.graphviz) {
          // Generate the GraphViz DOT representation using the existing utility
          const graphOutput = generateFiberDependencyGraph(
            fibersToShow,
            dependencyMap,
            reverseDependencyMap,
            detectionInfo,
            config as any,
            { reverse: options.reverse }
          );
          
          console.log(graphOutput);
          return;
        }
        
        // Sort fibers alphabetically for flat display, with core first
        const sortedFibers = orderFibers(fibersToShow, config, moduleResult.modules, {
          sortAlphabetically: true,
          handleSpecialFibers: true,
          includeDiscovered: false,
          hideDisabled: options.hideDisabled,
          reverse: options.reverse
        });
        
        if (options.flat) {
          console.log('Fiber dependencies:\n');
          
          for (const fiberId of sortedFibers) {
            const isCore = fiberId === FIBER_NAMES.CORE;
            const isEnabled = isCore || isFiberEnabled(fiberId, config);
            const statusSymbol = isEnabled ? EMOJI.ACTIVE : EMOJI.DISABLED;
            
            // Get explicit dependencies
            const deps = [...(activeMap.get(fiberId) || [])];
            
            // Add implicit core dependency for non-core fibers
            if (!isCore && fibersToShow.includes('core') && !deps.includes('core')) {
              deps.push('core');
            }
            
            // Sort dependencies alphabetically for consistent display
            const sortedDeps = [...deps].sort((a: string, b: string) => {
              if (a === 'core') return -1;
              if (b === 'core') return 1;
              return a.localeCompare(b);
            });
            
            const depsText = sortedDeps.length > 0 
              ? sortedDeps.join(', ')
              : '(none)';
              
            if (options.reverse) {
              console.log(`${statusSymbol} ${fiberId}: required by ${depsText}`);
            } else {
              console.log(`${statusSymbol} ${fiberId}: requires ${depsText}`);
            }
            
            if (options.detailed) {
              // Get dependency sources
              const sources = [...(detectionInfo.get(fiberId) || [])];
              
              // Add implicit core dependency source if not already present
              if (!isCore && fibersToShow.includes('core') && 
                  !sources.some(s => s.deps.includes('core'))) {
                sources.push({
                  source: 'implicit.core',
                  deps: ['core']
                });
              }
              
              if (sources.length > 0) {
                for (const source of sources) {
                  console.log(`   Source: ${source.source}`);
                  console.log(`   Dependencies: ${source.deps.join(', ')}`);
                }
              }
              console.log('');
            }
          }
        } else {
          // Tree view
          console.log('Fiber Dependency Diagram:');
          console.log('─────────────────────────');
          
          // We'll build a tree in the correct orientation
          const processedFibers = new Set<string>();
          
          // For reverse mode, we need to use the dependency map
          // For normal mode, create a reverse map to show what depends on each fiber
          const displayMap = options.reverse ? 
            dependencyMap :  // Reverse mode: what each fiber requires (directly from dependencyMap)
            new Map<string, string[]>(); // Normal mode: what depends on each fiber (built from dependencyMap)
            
          if (!options.reverse) {
            // Build the reversed map for normal display mode
            for (const [fiberId, deps] of dependencyMap.entries()) {
              for (const dep of deps) {
                if (!displayMap.has(dep)) {
                  displayMap.set(dep, []);
                }
                const dependents = displayMap.get(dep)!;
                if (!dependents.includes(fiberId)) {
                  dependents.push(fiberId);
                }
              }
            }
            
            // Remove redundant paths - if A depends on B and C, and B depends on C,
            // then we should only show A -> B -> C, not A -> C directly
            for (const [parent, dependents] of displayMap.entries()) {
              // Filter out paths where one dependent is also dependent on another in this list
              const directDependents = [...dependents];
              
              for (let i = dependents.length - 1; i >= 0; i--) {
                const dependent = dependents[i];
                
                // Check if this dependent depends on any other dependent
                for (const otherDependent of dependents) {
                  if (dependent !== otherDependent && 
                      (dependencyMap.get(dependent) || []).includes(otherDependent)) {
                    // Remove this path as it's redundant
                    const idx = directDependents.indexOf(dependent);
                    if (idx >= 0) {
                      directDependents.splice(idx, 1);
                    }
                    break;
                  }
                }
              }
              
              // Update with filtered list
              displayMap.set(parent, directDependents);
            }
            
            // Sort dependents in each entry putting dev first for core, etc.
            for (const [fiberId, dependents] of displayMap.entries()) {
              dependents.sort((a, b) => {
                if (fiberId === FIBER_NAMES.CORE) {
                  if (a === 'dev') return -1;
                  if (b === 'dev') return 1;
                  
                  if (a === 'dotfiles') return -1;
                  if (b === 'dotfiles') return 1;
                }
                return a.localeCompare(b);
              });
            }
          }
          
          // Function to print a node and its dependents
          function printDependencyTree(
            fiberId: string,
            prefix = '',
            isLast = true,
            depth = 0
          ) {
            // Skip if already processed (prevents cycles)
            if (processedFibers.has(fiberId)) {
              return;
            }
            
            // Mark as processed
            processedFibers.add(fiberId);
            
            // Print this node and its dependencies
            printDependencyNode(
              config,
              fiberId,
              prefix,
              isLast,
              options.detailed ? detectionInfo : null,
              fibersToShow
            );
            
            // Get and print dependents
            const dependents = displayMap.get(fiberId) || [];
            const newPrefix = `${prefix}${isLast ? DEPENDENCY_DISPLAY.TREE_INDENT : DEPENDENCY_DISPLAY.TREE_VERTICAL}`;
            
            // Print each dependent
            for (let i = 0; i < dependents.length; i++) {
              printDependencyTree(
                dependents[i],
                newPrefix,
                i === dependents.length - 1,
                depth + 1
              );
            }
          }
          
          // Start with core as the root in normal mode
          if (fibersToShow.includes(FIBER_NAMES.CORE)) {
            printDependencyTree(FIBER_NAMES.CORE);
          } else {
            // If core isn't available, find roots by dependency analysis
            const roots = options.reverse ?
              // In reverse mode, roots are fibers with no dependencies
              fibersToShow.filter(id => !dependencyMap.get(id)?.length) :
              // In normal mode, roots are fibers that nobody depends on
              fibersToShow.filter(id => !displayMap.get(id)?.length);
              
            // Sort roots alphabetically
            const sortedRoots = [...roots].sort((a, b) => a.localeCompare(b));
            
            // Print each root
            for (let i = 0; i < sortedRoots.length; i++) {
              printDependencyTree(sortedRoots[i], '', i === sortedRoots.length - 1);
            }
          }
        }
      } catch (error) {
        console.error('Error displaying fiber dependencies:', error);
        process.exit(1);
      }
    });
}

export function displayDependencyStatus(
  fiberId: string,
  isEnabled: boolean,
  isSatisfied: boolean,
  inConfig: boolean
): void {
  const statusSymbol = isEnabled ? EMOJI.ACTIVE : EMOJI.DISABLED;
  console.log(`  ${statusSymbol} ${fiberId}`);
}

export function displayChainDependencyStatus(
  chainId: string,
  isEnabled: boolean,
  isSatisfied: boolean,
  inConfig: boolean
): void {
  const statusSymbol = isEnabled ? EMOJI.ACTIVE : EMOJI.DISABLED;
  console.log(`  ${statusSymbol} ${chainId}`);
} 
