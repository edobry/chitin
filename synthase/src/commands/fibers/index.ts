import { Command } from 'commander';
import { serializeToYaml } from '../../utils';
import { loadAndValidateConfig } from '../utils';
import { discoverModulesFromConfig } from '../../modules/discovery';
import { validateModulesAgainstConfig } from '../../modules/validator';
import { UserConfig } from '../../config/types';
import { Module } from '../../modules/types';
import { 
  getLoadableFibers, 
  areFiberDependenciesSatisfied, 
  createChainFilter,
  getFiberIds,
  getChainIds,
  orderChainsByDependencies,
  generateFiberDependencyGraph
} from '../../fiber';

import {
  orderFibersByDependencies,
  getDependentFibers,
  isFiberEnabled,
  getChainDependencies,
  countDisplayedModules,
  ensureCoreDependencies
} from './utils';

import {
  displayFiberHeader,
  getFiberStatus,
  getFiberPath,
  displayValidationResults,
  displayFiberDependencies,
  displayChain,
  displaySummary,
  displayFiber
} from './display';

import {
  associateChainsByFiber,
  filterDisabledFibers,
  orderFibersByConfigAndName
} from './organization';

import { join } from 'path';
import fs from 'fs';
import yaml from 'js-yaml';
import { FIBER_NAMES } from '../../fiber/types';
import { CONFIG_FIELDS } from '../../config/types';
import { FILE_NAMES, EMOJI } from '../../constants';
import { createDepsCommand } from './deps-command';
import { createGetCommand } from './get-command';

/**
 * Find a module in a list of modules by ID and optionally by type
 * @param modules List of modules to search
 * @param id Module ID to find
 * @param type Optional module type filter
 * @returns The found module or undefined
 */
function findModuleById(modules: Module[], id: string, type?: 'fiber' | 'chain'): Module | undefined {
  return modules.find(module => 
    module.id === id && (type === undefined || module.type === type)
  );
}

/**
 * Shared function to load config and discover modules
 * @param options Command options
 * @returns Loaded configuration and module data
 */
async function loadConfigAndModules(options: any) {
  // Load and validate configuration
  const { config, validation } = await loadAndValidateConfig({
    userConfigPath: options.path,
    exitOnError: false
  });
  
  if (!validation.valid) {
    console.error('Configuration validation failed:');
    validation.errors.forEach(error => console.error(`- ${error}`));
    console.error('Continuing despite validation errors...');
  }
  
  // Discover and validate modules (silently)
  const moduleResult = await discoverModulesFromConfig(
    config, 
    options.baseDirs || []
  );
  
  if (moduleResult.errors.length > 0) {
    console.error('Encountered errors during module discovery:');
    for (const error of moduleResult.errors) {
      console.error(`- ${error}`);
    }
  }

  // Always run validation
  const validationResults = validateModulesAgainstConfig(
    moduleResult.modules,
    config
  );
  
  // Simple dependency checker for demonstration
  const dependencyChecker = (tool: string) => {
    if (options.detailed) console.error(`Checking dependency: ${tool} (assuming available)`);
    return true;
  };
  
  // Get all fibers from config
  const allFibers = getFiberIds(config);
  
  // Get loadable fibers
  const loadableFibers = options.available ? 
    getLoadableFibers(config, dependencyChecker) :
    allFibers;
  
  // Order fibers by dependency (foundational first, dependent last)
  const orderedFibers = orderFibersByDependencies(loadableFibers, config, moduleResult.modules);
  
  // Get all chains from all fibers
  const allChainIds = getChainIds(config);
  
  // Create chain filter based on loadable fibers
  const chainFilter = createChainFilter(config, loadableFibers);
  
  // Filter the chains that will be loaded
  const loadableChains = allChainIds.filter(chainId => chainFilter(chainId));
  
  // Order chains by dependencies
  const orderedChains = orderChainsByDependencies(loadableChains, config, loadableFibers);
  
  // Get all fiber modules from module discovery
  const discoveredFiberModules = moduleResult.modules.filter((m: Module) => m.type === 'fiber');
  
  // Create a map for all discovered fibers with their IDs as keys
  const discoveredFiberMap = new Map<string, Module>(
    discoveredFiberModules.map((module: Module) => [module.id, module])
  );
  
  // Get all chain modules from module discovery
  const discoveredChainModules = moduleResult.modules.filter((m: Module) => m.type === 'chain');
  
  // Create a map for all discovered chains with their IDs as keys
  const discoveredChainMap = new Map<string, Module>(
    discoveredChainModules.map((module: Module) => [module.id, module])
  );
  
  // Combine all fibers - configured and unconfigured for unified display
  const allFiberModuleIds = new Set([
    ...allFibers,
    ...discoveredFiberModules.map((m: Module) => m.id)
  ]);
  
  // Create a list of all fiber IDs to display ensuring proper dependency order
  // First, use the orderedFibers array to maintain dependency relationships
  let displayFiberIds = [...orderedFibers];
  
  // Add any discovered fibers that weren't in the ordered list
  Array.from(allFiberModuleIds)
    .filter(id => !displayFiberIds.includes(id))
    .sort((a, b) => a.localeCompare(b))
    .forEach(id => displayFiberIds.push(id));
  
  // Apply hide-disabled option if selected
  if (options.hideDisabled) {
    displayFiberIds = filterDisabledFibers(displayFiberIds, config, options.hideDisabled);
  }

  // Associate chains with their fibers
  const fiberChainMap = associateChainsByFiber(
    displayFiberIds,
    config,
    discoveredChainModules,
    discoveredFiberMap,
    moduleResult
  );
  
  return {
    config,
    validation,
    moduleResult,
    validationResults,
    allFibers,
    loadableFibers,
    displayFiberIds,
    discoveredFiberModules,
    discoveredFiberMap,
    discoveredChainModules,
    discoveredChainMap,
    orderedFibers,
    orderedChains,
    fiberChainMap,
    loadableChains,
    dependencyChecker
  };
}

/**
 * Create the 'list' subcommand for listing fiber names
 * @returns Configured Command object
 */
function createListCommand(): Command {
  return new Command('list')
    .description('List available fiber names')
    .option('-a, --available', 'List only available fibers (enabled with satisfied dependencies)')
    .option('-H, --hide-disabled', 'Hide disabled fibers')
    .option('-p, --path <path>', 'Custom path to user config file')
    .option('-b, --base-dirs <dirs...>', 'Additional base directories to scan for modules')
    .action(async (options) => {
      try {
        const {
          config,
          allFibers,
          loadableFibers,
          displayFiberIds,
          dependencyChecker
        } = await loadConfigAndModules(options);
        
        let fibersToList = options.available ? loadableFibers : displayFiberIds;
        
        // Apply hide-disabled option
        if (options.hideDisabled) {
          fibersToList = fibersToList.filter(fiberId => {
            const isCore = fiberId === 'core';
            return isCore || isFiberEnabled(fiberId, config);
          });
        }
        
        // Only output fiber names to stdout, one per line, with no headers or decorations
        for (const fiberId of fibersToList) {
          console.log(fiberId);
        }
      } catch (error) {
        console.error('Error listing fibers:', error);
        process.exit(1);
      }
    });
}

/**
 * Create the 'config' subcommand for showing fiber configuration
 * @returns Configured Command object
 */
function createConfigCommand(): Command {
  return new Command('config')
    .description('Display the configuration for a specific fiber')
    .argument('<name>', 'Name of the fiber to show configuration for')
    .option('-p, --path <path>', 'Custom path to user config file')
    .option('-j, --json', 'Output in JSON format instead of YAML')
    .action(async (name, options) => {
      try {
        const { config } = await loadConfigAndModules(options);
        
        // Check if the fiber exists in config
        if (!(name in config)) {
          console.error(`Fiber '${name}' not found in configuration.`);
          process.exit(1);
        }
        
        // Get the fiber's configuration
        const fiberConfig = config[name];
        
        // Output the configuration in YAML or JSON format
        if (options.json) {
          console.log(JSON.stringify(fiberConfig, null, 2));
        } else {
          console.log(serializeToYaml(fiberConfig as Record<string, unknown>));
        }
      } catch (error) {
        console.error('Error displaying fiber configuration:', error);
        process.exit(1);
      }
    });
}

/**
 * Create and configure the fibers command with subcommands
 * @returns Configured Command object
 */
export function createFibersCommand(): Command {
  const command = new Command('fibers')
    .description('Manage fibers and their modules')
    .addCommand(createGetCommand())
    .addCommand(createListCommand())
    .addCommand(createDepsCommand())
    .addCommand(createConfigCommand());

  // Make 'get' the default command when no subcommand is specified
  command.action(() => {
    command.help();
  });

  return command;
} 
