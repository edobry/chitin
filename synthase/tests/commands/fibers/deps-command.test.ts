#!/usr/bin/env bun
import { describe, test, expect, beforeEach, afterEach, spyOn } from 'bun:test';
import { createDepsCommand } from '../../../src/commands/fibers/deps-command';
import * as shared from '../../../src/commands/fibers/shared';
import { FIBER_NAMES } from '../../../src/fiber/types';
import { Module } from '../../../src/modules/types';
import { UserConfig, CONFIG_FIELDS } from '../../../src/config/types';

/**
 * Helper to create fiber modules for testing
 */
function createTestFiberModule(id: string, config: any = {}): Module {
  return {
    id,
    type: 'fiber',
    name: `${id.charAt(0).toUpperCase() + id.slice(1)} Fiber`,
    path: `/path/to/${id}`,
    config,
    metadata: {}
  } as Module;
}

/**
 * Helper to create a basic test environment with a simple dependency structure
 */
function createBasicTestEnvironment() {
  // Create a simple dependency structure: core <- dev <- app
  const config: UserConfig = {
    core: { enabled: true },
    dev: { enabled: true, fiberDeps: ['core'] },
    app: { enabled: true, fiberDeps: ['dev'] },
    utils: { enabled: true, fiberDeps: ['core'] },
    plugin: { enabled: false, fiberDeps: ['app'] }
  } as any;

  const modules = [
    createTestFiberModule('core'),
    createTestFiberModule('dev', { fiberDeps: ['core'] }),
    createTestFiberModule('app', { fiberDeps: ['dev'] }),
    createTestFiberModule('utils', { fiberDeps: ['core'] }),
    createTestFiberModule('plugin', { fiberDeps: ['app'] })
  ];

  const displayFiberIds = ['core', 'dev', 'app', 'utils', 'plugin'];
  const orderedFibers = ['core', 'dev', 'app', 'utils', 'plugin'];

  return {
    config,
    moduleResult: { modules },
    displayFiberIds,
    orderedFibers
  };
}

/**
 * Helper to create a test environment with more complex relationships
 */
function createComplexTestEnvironment() {
  const config: UserConfig = {
    [FIBER_NAMES.CORE]: { enabled: true },
    dev: { 
      enabled: true, 
      fiberDeps: [FIBER_NAMES.CORE],
      provides: ['dev-tool']
    },
    ui: {
      enabled: true,
      fiberDeps: ['dev'],
      provides: ['ui-component']
    },
    api: {
      enabled: true,
      fiberDeps: ['dev'],
      toolDeps: ['ui-component']
    },
    data: {
      enabled: false,
      fiberDeps: ['api', 'ui']
    }
  } as any;

  const modules = [
    createTestFiberModule(FIBER_NAMES.CORE),
    createTestFiberModule('dev', { fiberDeps: [FIBER_NAMES.CORE] }),
    createTestFiberModule('ui', { fiberDeps: ['dev'] }),
    createTestFiberModule('api', { fiberDeps: ['dev'], toolDeps: ['ui-component'] }),
    createTestFiberModule('data', { fiberDeps: ['api', 'ui'] })
  ];

  const displayFiberIds = [FIBER_NAMES.CORE, 'dev', 'ui', 'api', 'data'];
  const orderedFibers = [FIBER_NAMES.CORE, 'dev', 'ui', 'api', 'data'];

  return {
    config,
    moduleResult: { modules },
    displayFiberIds,
    orderedFibers
  };
}

describe('deps command', () => {
  // Mock console.log to capture output
  const originalConsoleLog = console.log;
  let consoleOutput: string[] = [];

  const mockConsoleLog = (...args: any[]) => {
    // Join arrays in args for a cleaner snapshot
    const processedArgs = args.map(arg => 
      typeof arg === 'object' && arg !== null ? JSON.stringify(arg, null, 2) : arg
    );
    consoleOutput.push(processedArgs.join(" "));
  };

  beforeEach(() => {
    // Reset console output capture
    consoleOutput = [];
    console.log = mockConsoleLog;
  });
  
  afterEach(() => {
    // Restore console.log
    console.log = originalConsoleLog;
  });

  // Basic command visualization tests
  describe('basic dependency structure', () => {
    beforeEach(() => {
      const env = createBasicTestEnvironment();
      spyOn(shared, 'loadConfigAndModules').mockResolvedValue(env);
    });

    test('should display tree diagram by default', async () => {
      const command = createDepsCommand();
      await command.parseAsync(["deps"]);
      
      // Verify we have some output
      expect(consoleOutput.length).toBeGreaterThan(0);
      
      // Snapshot test to catch unexpected changes in output format
      expect(consoleOutput.join("\n")).toMatchSnapshot();
    });
    
    test('should output JSON when --json flag is used', async () => {
      const command = createDepsCommand();
      await command.parseAsync(["deps", "--json"]);
      
      // Just check that we got some output
      const output = consoleOutput.join("\n");
      expect(output.length).toBeGreaterThan(0);
      
      // Snapshot test for JSON output
      expect(output).toMatchSnapshot();
    });
    
    test('should hide disabled fibers when --hide-disabled flag is used', async () => {
      const command = createDepsCommand();
      await command.parseAsync(["deps", "--hide-disabled"]);
      
      // Snapshot test for filtered output
      expect(consoleOutput.join("\n")).toMatchSnapshot();
    });
    
    test('should show flat list when --flat flag is used', async () => {
      const command = createDepsCommand();
      await command.parseAsync(["deps", "--flat"]);
      
      // Snapshot test for flat list
      expect(consoleOutput.join("\n")).toMatchSnapshot();
    });
  });

  // Complex dependency structure tests
  describe('complex dependency structure', () => {
    beforeEach(() => {
      const env = createComplexTestEnvironment();
      spyOn(shared, 'loadConfigAndModules').mockResolvedValue(env);
    });
    
    test('should show reverse dependencies when --reverse flag is used', async () => {
      const command = createDepsCommand();
      await command.parseAsync(["deps", "--reverse"]);
      
      // Snapshot test for reverse dependencies
      expect(consoleOutput.join("\n")).toMatchSnapshot();
    });
    
    test('should show detailed information when --detailed flag is used', async () => {
      const command = createDepsCommand();
      await command.parseAsync(["deps", "--detailed"]);
      
      // Snapshot test for detailed mode
      expect(consoleOutput.join("\n")).toMatchSnapshot();
    });
    
    test('should output GraphViz format when --graphviz flag is used', async () => {
      const command = createDepsCommand();
      await command.parseAsync(["deps", "--graphviz"]);
      
      // Just check we got some output
      const output = consoleOutput.join("\n");
      expect(output.length).toBeGreaterThan(0);
      
      // Snapshot test for GraphViz output
      expect(output).toMatchSnapshot();
    });
    
    test('should handle multiple flags together', async () => {
      const command = createDepsCommand();
      await command.parseAsync(["deps", "--hide-disabled", "--reverse", "--flat"]);
      
      // Snapshot test for combined flags
      expect(consoleOutput.join("\n")).toMatchSnapshot();
    });
  });
}); 
