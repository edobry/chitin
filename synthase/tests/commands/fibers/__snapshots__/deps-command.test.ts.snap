// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`deps command basic dependency structure should display tree diagram by default 1`] = `
"Fiber Dependency Diagram:
â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€
â””â”€â”€ ðŸŸ¢ core
    â”œâ”€â”€ ðŸŸ¢ dev
    â”‚   â””â”€â”€ ðŸŸ¢ app
    â”‚       â””â”€â”€ âš« plugin
    â””â”€â”€ ðŸŸ¢ utils"
`;

exports[`deps command basic dependency structure should output JSON when --json flag is used 1`] = `
"Fiber Dependency Diagram:
â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€
â””â”€â”€ ðŸŸ¢ core
    â”œâ”€â”€ ðŸŸ¢ dev
    â”‚   â””â”€â”€ ðŸŸ¢ app
    â”‚       â””â”€â”€ âš« plugin
    â””â”€â”€ ðŸŸ¢ utils"
`;

exports[`deps command basic dependency structure should hide disabled fibers when --hide-disabled flag is used 1`] = `
"Fiber Dependency Diagram:
â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€
â””â”€â”€ ðŸŸ¢ core
    â”œâ”€â”€ ðŸŸ¢ dev
    â”‚   â””â”€â”€ ðŸŸ¢ app
    â”‚       â””â”€â”€ âš« plugin
    â””â”€â”€ ðŸŸ¢ utils"
`;

exports[`deps command basic dependency structure should show flat list when --flat flag is used 1`] = `
"Fiber Dependency Diagram:
â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€
â””â”€â”€ ðŸŸ¢ core
    â”œâ”€â”€ ðŸŸ¢ dev
    â”‚   â””â”€â”€ ðŸŸ¢ app
    â”‚       â””â”€â”€ âš« plugin
    â””â”€â”€ ðŸŸ¢ utils"
`;

exports[`deps command complex dependency structure should show reverse dependencies when --reverse flag is used 1`] = `
"Fiber Dependency Diagram:
â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€
â””â”€â”€ ðŸŸ¢ core
    â””â”€â”€ ðŸŸ¢ dev
        â””â”€â”€ ðŸŸ¢ ui
            â””â”€â”€ ðŸŸ¢ api
                â””â”€â”€ âš« data"
`;

exports[`deps command complex dependency structure should show detailed information when --detailed flag is used 1`] = `
"Fiber Dependency Diagram:
â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€
â””â”€â”€ ðŸŸ¢ core
    â””â”€â”€ ðŸŸ¢ dev
        â””â”€â”€ ðŸŸ¢ ui
            â””â”€â”€ ðŸŸ¢ api
                â””â”€â”€ âš« data"
`;

exports[`deps command complex dependency structure should output GraphViz format when --graphviz flag is used 1`] = `
"Fiber Dependency Diagram:
â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€
â””â”€â”€ ðŸŸ¢ core
    â””â”€â”€ ðŸŸ¢ dev
        â””â”€â”€ ðŸŸ¢ ui
            â””â”€â”€ ðŸŸ¢ api
                â””â”€â”€ âš« data"
`;

exports[`deps command complex dependency structure should handle multiple flags together 1`] = `
"Fiber dependencies:

ðŸŸ¢ core: required by api, data, dev, ui
ðŸŸ¢ api: required by core, data
âš« data: required by core
ðŸŸ¢ dev: required by core, api, ui
ðŸŸ¢ ui: required by core, api, data"
`;
